examples/shape.rs

```diff
@@ -0,0 +1,344 @@
+
+        let hhea = otf
+            .read_table(&file.scope, tag::HHEA)?
+            .ok_or_else(|| anyhow!("HHEA table not found"))?
+            .read::<HheaTable>()?;
+        let hmtx = otf
+            .read_table(&file.scope, tag::HMTX)?
+            .ok_or_else(|| anyhow!("HMTX table not found"))?
+            .read_dep::<HmtxTable>((
+                usize::from(maxp.num_glyphs),
+                usize::from(hhea.num_h_metrics),
+            ))?;
+
+        let gsub_table = otf
+            .find_table_record(tag::GSUB)
+            .ok_or_else(|| anyhow!("GSUB table record not found"))?
+            .read_table(&file.scope)?
+            .read::<LayoutTable<GSUB>>()?;
+        let gdef_table: Option<GDEFTable> = otf
+            .find_table_record(tag::GDEF)
+            .map(|gdef_record| -> Result<GDEFTable> {
+                Ok(gdef_record.read_table(&file.scope)?.read::<GDEFTable>()?)
+            })
+            .transpose()?;
+        let opt_gpos_table = otf
+            .find_table_record(tag::GPOS)
+            .map(|gpos_record| -> Result<LayoutTable<GPOS>> {
+                Ok(gpos_record
+                    .read_table(&file.scope)?
+                    .read::<LayoutTable<GPOS>>()?)
+            })
+            .transpose()?;
+        let gsub_cache = new_layout_cache(gsub_table);
+        let gpos_cache = opt_gpos_table.map(new_layout_cache);
+
+        Ok(Self {
+            cmap_subtable,
+            hmtx,
+            hhea,
+            gpos_cache,
+            gsub_cache,
+            gdef_table,
+            num_glyphs,
+            _scope: owned_scope,
+        })
+    }
+
+    fn glyph_index_for_char(&self, c: char) -> Result<Option<u16>> {
+        self.cmap_subtable
+            .map_glyph(c as u32)
+            .map_err(|e| anyhow!("Error while looking up glyph {}: {}", c, e))
+    }
+
+    pub fn shape_text<T: AsRef<str>>(
+        &self,
+        text: T,
+        font_index: usize,
+        script: u32,
+        lang: u32,
+    ) -> Result<Vec<Shaped>> {
+        let mut glyphs = vec![];
+        for c in text.as_ref().chars() {
+            glyphs.push(RawGlyph {
+                unicodes: vec![c],
+                glyph_index: self.glyph_index_for_char(c)?,
+                liga_component_pos: 0,
+                glyph_origin: GlyphOrigin::Char(c),
+                small_caps: false,
+                multi_subst_dup: false,
+                is_vert_alt: false,
+                fake_bold: false,
+                fake_italic: false,
+                extra_data: (),
+            });
+        }
+
+        let vertical = false;
+
+        gsub_apply_default(
+            &|| vec![], //map_char('\u{25cc}')],
+            &self.gsub_cache,
+            self.gdef_table.as_ref(),
+            script,
+            lang,
+            vertical,
+            self.num_glyphs,
+            &mut glyphs,
+        )?;
+
+        // Note: init_from_glyphs silently elides entries that
```

Comment:
This behavior was surprising; it would be great if there was a reasonable way to signal that the glyph was unknown rather than stripping it out here:

https://docs.rs/allsorts/0.1.0/src/allsorts/gpos.rs.html#292